### ISSUE TITLE
How to build on top of this?

### ISSUE THREAD

**Issue by nothing2code** [phil-opp/blog_os#1341]
Can I expand upon the foundation I've built through this blog?

I have a few specific goals in mind for further development:
- Implementing a framebuffer
- Building a simple terminal to execute commands
- Attempting to run Doom (which I realize might be more challenging than I anticipate)

Is it feasible to achieve these with the groundwork I've laid by following this blog? Additionally, where can I find the necessary resources to guide me through these next steps?

**Comment by ViktorPopp** [comment]
> ## Implementing a frame buffer
> Just use edition 3, the bootloader crates give you a frame buffer. ez

Not quite. Some parts of the paging needs ti be rewritten.

**Comment by tsatke** [comment]
@ChocolateLoverRaj to port doom, you set up a `gcc` toolchain that works with your os and compile doom from source, then fix all the problems you encounter. I'd assume that your os would need to be POSIX compatible for that, I'm not sure, as I've never attempted it.

**Comment by ViktorPopp** [comment]
@tsatke it is easier just creating a linux/posix compatible elf and syscall abi

**Comment by tsatke** [comment]
Can you rephrase the question to be more precise?

Easier than what?

> create an elf abi

You can create your own formats based on elf (apple did this for example), but that's not needed here

> create a Linux compatible syscall abi 

In case you actually mean ABI, use C or sysv64, no need to create your own. Don't know what Linux uses, but I'd assume one of those. 

> create a posix compatible syscall abi

In case you actually mean ABI? Posix doesn't care for all I know. Your syscall abi is between your Kernel and your OS. 

In case you mean syscalls, posix specifies them for you. This is stuff like `read`, `fopen` or `mmap` that you know from for example the C stdlib.

**Comment by bjorn3** [comment]
> You can create your own formats based on elf (apple did this for example)

Mach-O and ELF are entirely unrelated. They were created roughly at the same time and have a quite different structure.

**Comment by ViktorPopp** [comment]
Sorry for the misunderstanding. I meant creating a Linux compatible syscall ABI and an elf loader (or elf compatible). Then DOOM and other applications can be compiled for linux and run natively on your own OS.
